File: backend/app/domain/tags.py
from __future__ import annotations

import re
from collections.abc import Iterable


TAG_NAME_PATTERN = re.compile(r"^[0-9A-Za-zА-Яа-яЁё_]+$")


def normalize_tag_name(value: str) -> str:
    normalized = value.strip().lower()
    if not normalized:
        raise ValueError("Tag name cannot be empty")
    if any(character.isspace() for character in normalized):
        raise ValueError("Tag name cannot contain spaces")
    if not TAG_NAME_PATTERN.fullmatch(normalized):
        raise ValueError(
            "Tag name may contain only Latin/Cyrillic letters, digits, and underscore"
        )
    return normalized


def normalize_tag_names(values: Iterable[str]) -> list[str]:
    normalized_values: list[str] = []
    seen: set[str] = set()

    for value in values:
        normalized = normalize_tag_name(value)
        if normalized in seen:
            continue
        seen.add(normalized)
        normalized_values.append(normalized)

    return normalized_values

File: frontend/src/tags.ts
const TAG_NAME_PATTERN = /^[0-9A-Za-zА-Яа-яЁё_]+$/

export function normalizeTagName(value: string): string {
  const normalized = value.trim().toLowerCase()

  if (!normalized) {
    throw new Error("Tag name cannot be empty")
  }

  if (/\s/.test(normalized)) {
    throw new Error("Tag name cannot contain spaces")
  }

  if (!TAG_NAME_PATTERN.test(normalized)) {
    throw new Error("Tag name may contain only Latin/Cyrillic letters, digits, and underscore")
  }

  return normalized
}

export function normalizeTagNames(values: string[]): string[] {
  const seen = new Set<string>()
  const normalizedValues: string[] = []

  for (const value of values) {
    const normalized = normalizeTagName(value)

    if (seen.has(normalized)) {
      continue
    }

    seen.add(normalized)
    normalizedValues.push(normalized)
  }

  return normalizedValues
}
